lib/distributed_cache.rb

```diff
@@ -12,3 +12,52 @@ def initialize(key, manager: nil, namespace: true)
     )
   end
 end
+
+module HasDistributedCache
+  module Test
+    module ClassMethods
+      def caches
+        @caches ||= {}
+      end
+
+      def dirty
+        @dirty ||= Set.new
+      end
+
+      def clear_caches!
+        dirty.each do |object_id|
+          caches[object_id].each_value do |cache|
+            cache.clear
+          end
+        end
+      end
+    end
+
+    def self.included(base)
+      base.extend ClassMethods
+    end
+
+    def has_distributed_cache(name, key, **opts)
+      define_singleton_method(name) do
+        HasDistributedCache.dirty.add(object_id)
+        HasDistributedCache.caches[object_id] ||= {}
```

Comment:
if you simply set HasDistributedCache.caches to `{}` at the end of every test run no need to track dirty ? no?

Reply:
@SamSaffron, This comment was extremely helpful.

I had the same thought yesterday, but I could see a clear performance regression when I did that. I reasoned, falsely, that there must be an important difference around when caches are initialized.

When I revisited it this morning, with a fresh brain, that explanation didn't hold water for me. It turns out that the performance difference is because the dirty tracking version has an important bug which actually prevents it from working. The extra ~40 seconds isn't down to redis traffic or cache initialization at all. It's just that the tests use cached values from prior tests.

Here's the data. This is cache hits by cache where the value was written in a different test:

```
["icon_manager", 10854]
["theme", 6829]
["am_serializer_fragment_cache", 4840]
["banner_json", 1299]
["scheme_hex_for_name", 1263]
["discourse_stylesheet", 311]
["category_topic_ids", 148]
["svg_sprite", 66]
["csp_extensions", 12]
["developer_ids", 4]
["last_read_only", 0]
["category_url", 0]
```

When you give each test a clear cache, it rebuilds it by doing queries against the DB. It seems obvious in retrospect, but these things often do.

My plan is to prime these caches explicitly before each test.